src/include/miopen/tensor.hpp

```diff
@@ -175,37 +176,50 @@ struct TensorDescriptor : miopenTensorDescriptor
 
     std::string ToString() const;
 
-    template <class Vector, class Op>
-    static inline std::vector<int64_t> sort_permutation(const Vector& data, Op op)
+    std::string GetLayout(std::string labels) const
     {
-        std::vector<std::int64_t> result(data.size());
-        std::iota(result.begin(), result.end(), 0);
-        std::sort(
-            result.begin(), result.end(), [&](auto x, auto y) { return op(data[x], data[y]); });
-        return result;
+        std::ignore = labels;
+        return layout;
     }
 
-    std::string GetLayout(std::string labels) const
+    std::string ComputeLayout(const std::string& labels)
     {
-        if(labels.size() != strides.size())
+        //std::cout << "lens: ";
+        //printVector(lens);
+        //std::cout << "Strides: ";
+        //printVector(strides);
+
+        auto valid_layouts = compute_valid_layouts(strides, lens, labels);
```

Comment:
I dont think we need to compute all layouts. We can compute the permutation correctly by minimizing the adjacent difference. This can be done by grouping the duplicate strides and then sort them based on the difference from the previous index:

```cpp
template <class Iterator, class Output, class Predicate>
void group_unique(Iterator start, Iterator last, Output out, Predicate pred)
{
    while(start != last)
    {
        auto it = std::find_if(start, last, [&](auto&& x) { return not pred(*start, x); });
        out(start, it);
        start = it;
    }
}

std::vector<int64_t> find_permutation(const std::vector<std::size_t>& strides)
{
    std::vector<std::int64_t> result;
    std::multimap<std::size_t, int64_t, std::greater<>> m;
    int i = 0;
    for(auto&& x : strides)
        m.emplace(x, i++);
    auto pred     = by(std::equal_to<>{}, [](auto&& p) { return p.first; });
    int64_t base  = 0;
    int64_t delta = 0;
    auto each     = [&](auto start, auto last) {
        auto n = std::distance(start, last);
        std::transform(start, last, std::back_inserter(result), [](auto&& p) { return p.second; });
        auto mag = delta > 0 ? -1 : 1;
        std::stable_sort(result.end() - n, result.end(), by(std::less<>{}, [&](auto x) {
                             return (x - base) * mag;
                         }));
        delta = result.back() - base;
        base  = result.back();
    };
    group_unique(m.begin(), m.end(), each, pred);
    return result;
}
```

Reply:
Although this will still be ambiguous for cases like `64, 1, 1, 1`.

Reply:
> auto pred     = by(std::equal_to<>{}, [](auto&& p) { return p.first; });

@pfultz2 Just for understanding what does this `by` keyword do?

> then sort them based on the difference from the previous index

The original index is based `NCHW`, why is `NCHW` the golden standard to judge all other layouts?

Reply:
> Just for understanding what does this by keyword do?

It applies projection to an operator. It is defined [here](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/src/include/miopen/functional.hpp#L46). So `by(f, p)(x, y)` is equivalent to `f(p(x), p(y))`.

> The original index is based NCHW, why is NCHW the golden standard to judge all other layouts?

Because that is the order of dimensions regardless of layout in memory.

Reply:
I still don't have a good enough understanding of your algorithm.

> auto it = std::find_if(start, last, [&](auto&& x) { return not pred(*start, x); });

`pred()` returns true if strides are equal, but with a not in front you will find those cases whose strides are different. I'd thought you would do otherwise and only calculate the adjacent difference for strides equal cases because those are ambiguous indexes.

>    int64_t base  = 0;
>    int64_t delta = 0;
>    auto each     = [&](auto start, auto last) { ... }

I got completely lost in what it is trying to do in this lambda. I was trying to come up with a minimal example but can't get it to compile, [here](http://cpp.sh/83qpz). Could you help with making it to compile, and click on "GetUrl", just so that I can step through for better understanding.

I do realize that it doesn't try to check the calculated layout's validity like:
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/5203a870b3f27d91750e17d10112565545138b60/src/include/miopen/tensor_layout.hpp#L179-L188

I think this is still essential to include to confirm that derived strides are consistent with lens.

Reply:
Either way, I don't think this thread necessarily block the PR. Could you take a look at 31e693f? If we agree this is the way to go, I can revert the previous two commits and have carlus unblocked.

Reply:
> pred() returns true if strides are equal, but with a not in front you will find those cases whose strides are different.

This is part of the grouping, its trying to find the next element that is not equal, so it can find the end to the subrange. This grouping is done since there is not a way to traverse just the keys. So instead it iterates over the elements `(x, a), (y, b), (y, c), (z, d)` and tries to group them into 3 subranges: `{(x, a)}, {(y, b), (y, c)}, {(z, d)}`. It passes the iterators to the `each` lambda for each subrange, where `last` is one past the end(just like calling `.end()` on STL containers). 

Of course this only works when duplicate keys appear together(which it will with `std::multimap`). That is why it is called `group_unique`(instead of `group_by`) as `std::unique` relies on the same precondition. To make it work for unordered containers then `std::partition` can be used instead(but `std::partition` wont work for associative containers). You can see how that is implemented in migraphx's [`group_by`](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/src/include/migraphx/algorithm.hpp#L11).

So once it has everything grouped together it tries to sort the values based on the delta of the previous in order to minimize the adjacent difference and direction in the permutation. So if `strides[1] == strides[3]` then we could have two valid permutations: `0,2,1,3` and `0,2,3,1`. So in this case, the delta of 2 with 3 or 1 is the same, but the direction is different. The permutation is increasing in value so `0,2,3,1` should be chosen instead.

> I think it doesn't try to check the calculated layout's validity like below:

No it doesn't but I think the sort comparator could be updated to take in account the lens so it can sort the group better, rather than just minimize adjacent difference. This is even simpler as we don't need to put it in a multimap or grouping:

```cpp
std::vector<int64_t> find_permutation(const std::vector<std::size_t>& lens, const std::vector<std::size_t>& strides)
{
    std::vector<std::int64_t> result(lens.size());
    std::iota(result.begin(), result.end(), 0);
    std::stable_sort(result.begin(), result.end(), by(std::greater<>{}, [&](auto x) {
                         return std::make_tuple(strides[x], lens[x]);
                     }));
    return result;
}
```

My initial tests show that this works, but it may need more testing or review to understand if this always produces a "valid" layout.

> I believe this is still essential to include to confirm that derived strides are consistent with lens.

But that wont work with non-packed tensors(which the tensor ops support), it should probably be an assert only when the tensor is packed.

Reply:
So thinking about this, I believe sorting with lens is correct, because the 1 lens should be moved to the end. In general, lens that share the same strides should all be ones except for one of them. The stride with a non-1 dimension should come first.

I dont think there would be scenario where you would have more than one dimension be non-1 that have the same strides. If that is the case we could change the comparator for sort to do `return std::make_tuple(strides[x], std::max<std::size_t>(2, lens[x]));`, which makes the dimension either 1 or 2 when sorting, which will ensure that `std::stable_sort` will use the order of dimensions in the tensor for the non-1 dimensions. However, this may not even be a possible scenario.

Reply:
Thanks a lot for the thorough explanation. I've got to say you are one step ahead of me in deriving this simple and elegant solution.

#### Adjacent difference solution
Counter example: 
lens = [16, 1, 3, 3]
strides = [1, 1, 48, 16]
The only valid layout is `HWNC`, however the adjacent difference solution will give a `HWCN`

#### Sorting lens && strides together

> The stride with a non-1 dimension should come first.

Excellent. Once this requirement get clarified, you solution makes total sense. The part that I like this solution in that even for the cases where we got multiple 1s, we are guaranteed to give back the valid solution that always start with the non-1 dimension. `[1, 1, 1, 1]` strides actually give back the perfect `NCHW` result.